WikiWatershed/model-my-watershed
# Manual Entry: Recalculate Area-derived Values on Land Cover Updates

## Overview

MapShed operates on a large dictionary of values, whose fields can have static, lookup, user-provided, or calculated values. A number of these values are calculated using rasters geoprocessed for the given area of interest. One of these rasters in the NLCD raster, which is provides a count of cells for each land cover type. These counts are stored in the `Area` key, but are also used to calculate some derived values. The final dictionary with all the values is saved to the `gis_data` field of the project.

As of #2977, users can override the land cover counts for a given scenario. This modification is saved on the scenario. When running MapShed, these modifications are combined with the `gis_data` of the project to produce a scenario-specific dictionary, whose values are used to run GWLF-E and get results.

Unfortunately, so far we have been simply overriding the `Area` key of `gis_data`, and not any of the other keys which derive their value from the `Area` list. This is partially addressed in this PR.

Keys whose values are affected by the `Area` list are extracted into an `area_calculations` function which is called from `collect_data`, and also from `apply_gwlfe_modifications`, to ensure that their values are correctly derived on initialization and update.

This is done on the back-end for facilitating improved GWLF-E calculations, and on the front-end for allowing Land Cover modifications to inform Conservation Practice modifications.

See commit messages for details.

Connects #3142 

### Demo

```console
$ ./scripts/manage.sh test_mapshed
+ ARGS=test_mapshed
+ [[ 1 -eq 1 ]]
+ [[ test_mapshed == runserver ]]
+ vagrant ssh app -c 'cd /opt/app && envdir /etc/mmw.d/env ./manage.py test_mapshed'
.
----------------------------------------------------------------------
Ran 1 test in 1.124s

OK
Connection to 127.0.0.1 closed.
```

If the land use is overridden:

![localhost_8000_project_13_scenario_17_](https://user-images.githubusercontent.com/1430060/63364645-07f1f280-c344-11e9-8e64-e6e71affea01.png)

The conservation practice also updates:

![localhost_8000_project_13_scenario_17_ (1)](https://user-images.githubusercontent.com/1430060/63364656-0cb6a680-c344-11e9-80ad-e760cad87760.png)

### Notes

There are two exceptions to the fields updated, one overt and one subtle:

Despite being computed from `Area`, the value of `LS` is not recalculated, because it also relies on the `lu_stream_pct` geoprocessing result. This is not currently not saved, and thus cannot be used during recalculation. Thus, this field is left to its initially calculated value.

A number of other geoprocessing operations combine NLCD with another raster to get a grouped count, such as `nlcd_soil`, `nlcd_streams`, `nlcd_slope`, and `nlcd_kfactor`. These are in turn used to calculate a number of results, such as `cn`, `ag_stream_pct`, `low_urban_stream_pct`, `med_high_urban_stream_pct`, `lu_stream_pct`, `ag_slope_3_pct`, `ag_slope_3_8_pct`, `n41`, `kf`, and `avg_kf`. These are in turn used to calculate a number of MapShed fields, such as `AgLength`, `UrbLength`, `n42`, `n46e`, `n46f`, `CN`, `AgSlope3`, `AgSlope3To8`, `n41`, `AvKF`, `KF`, and `P`.

Recalculating these fields from the modifications is complicated, because the modifications are only changing the proportion of land use distribution, not the actual distribution in space. Without the spatial updates, we cannot run raster operations. This will need to be clarified and addressed at a later date.

Additionally, there are some concerns [outlined here](https://github.com/WikiWatershed/model-my-watershed/issues/3142#issuecomment-523088313) about the implications of modifications affecting each other.

## Testing Instructions

* Check out this branch and reprovision `app`

    ```console
    $ vagrant reload app --provision
    ```
* Check out only the first commit, which adds the test but doesn't contain the refactoring, and run the test:

    ```console
    $ git checkout HEAD~2
    $ ./scripts/manage.sh test_mapshed
    ```

* Check out the main branch, and run the test again to ensure it runs correctly:

    ```console
    $ git checkout -
    $ ./scripts/manage.sh test_mapshed
    ```

* Go to [:8000/](http://localhost:8000/) and create a MapShed project
* Add changes to Current Conditions
* Open the Land Cover modal
* Change the area of Cropland. Change the area of something else to compensate. Save your changes.
* Try and add a conservation practice, such as Cover Crops
  - [ ] Ensure the Area of Row Crops matches your user input value
* Go back to the Land Cover modal and reset to original values
* Try and add the same conservation practice again
  - [ ] Ensure the Area of Row Crops matches the original value

## Checklist
